client/clientservice/src/state_snapshot_service.cpp

```diff
@@ -22,27 +28,388 @@ using vmware::concord::client::statesnapshot::v1::StreamSnapshotResponse;
+        return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Invalid argument");
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::NOT_READY)):
+        LOG_INFO(logger, "Request failed with NOT_READY error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "No clients connected to the replicas");
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::TIMEOUT)):
+        LOG_INFO(logger, "Request failed with TIMEOUT error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "Timeout");
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::EXEC_DATA_TOO_LARGE)):
+        LOG_INFO(logger, "Request failed with EXEC_DATA_TOO_LARGE error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Execution data too large");
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::EXEC_DATA_EMPTY)):
+        LOG_INFO(logger, "Request failed with EXEC_DATA_EMPTY error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Execution data is empty");
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::CONFLICT_DETECTED)):
+        LOG_INFO(logger, "Request failed with CONFLICT_DETECTED error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "Aborted");
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::OVERLOADED)):
+        LOG_INFO(logger, "Request failed with OVERLOADED error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "All clients occupied");
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::UNKNOWN)):
+        LOG_INFO(logger, "Request failed with UNKNOWN error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Failure due to unknown reason");
+        break;
+      case (static_cast<uint32_t>(bftEngine::OperationResult::INTERNAL_ERROR)):
+        LOG_INFO(logger, "Request failed with INTERNAL error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Failure due to some internal error");
+        break;
+      default:
+        LOG_INFO(logger, "Request failed with INTERNAL error for cid=" << correlation_id);
+        return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Internal error");
+        break;
+    }
+    return;
+  }
+  return_status = grpc::Status::OK;
+  auto reply = std::get<bft::client::Reply>(send_result);
+  if (is_rsi) {
+    ReconfigurationResponse res;
+    concord::messages::deserialize(reply.matched_data, res);
+    response.response.template emplace<ResponseT>(std::get<ResponseT>(res.response));
```

Comment:
Interesting, I haven't seen this before. I assume you're calling `emplace` on the variant but `template` looks odd to me. How does this work?

Reply:
Hey @robem, As this is a template function all the types are not deduced. So we use template type computation here to call the correct emplace. You need to read template metaprogramming or template functional programming book (i don't remember the exact name). This concept should also be available on the internet and in courses of advanced C++.